src/libs/actions/App.js

```diff
@@ -116,45 +110,49 @@ AppState.addEventListener('change', (nextAppState) => {
  * Fetches data needed for app initialization
  */
 function openApp() {
-    API.read('OpenApp', {policyIDList}, {
-        optimisticData: [{
-            onyxMethod: CONST.ONYX.METHOD.MERGE,
-            key: ONYXKEYS.IS_LOADING_REPORT_DATA,
-            value: true,
-        }],
-        successData: [{
-            onyxMethod: CONST.ONYX.METHOD.MERGE,
-            key: ONYXKEYS.IS_LOADING_REPORT_DATA,
-            value: false,
-        }],
-        failureData: [{
-            onyxMethod: CONST.ONYX.METHOD.MERGE,
-            key: ONYXKEYS.IS_LOADING_REPORT_DATA,
-            value: false,
-        }],
+    policyIDListExcludingWorkspacesCreatedOfflineLoaded.then(() => {
```

Comment:
If we want to guarantee some values are read from storage in order for the App's init logic to run correctly this is certainly one way to do it. This could be a good opportunity to think a bit more about the "cold cache" problem that [we've run into before](https://github.com/Expensify/App/blob/e5281fcb343b9f9825711b871efb8909ea0df308/src/libs/Network/SequentialQueue.js#L62-L73) and see if we can't solve it another way.

The pattern we are introducing here looks a bit like what we were trying to get away from with the new API stuff in general (promise chaining) and is kind of a thinly veiled `Onyx.get()` (which we have avoided so far)..

I think if we are going to "wait for data from the cache" then we should either:

- accept there is a valid use case for waiting for data values to be read and expose `Onyx.get()` (not sure if there really is yet)
- implement some kind of "initialKeysToCache" array in Onyx so that the data we need when the app starts are already in the cache (would that solve this problem? or could `OpenApp` still run before a call to `Onyx.connect()` initializes the value of the policies).
- some other solution?

cc @tgolen

Reply:
Would `initialKeysToCache` be different than the existing param for `initialKeyStates`?

However, I'm not really sure that this would solve the problem that this PR is intending to solve, or if it does, I don't see how it solve the problem in this PR (since you can't know the policy IDs to seed the cache with).

I definitely agree with @marcaaron that this solution opens up a very slippery slope with chaining.

It seems to me like the policies will definitely be read from Onyx when `reconnectApp()` is called. Is that a correct assumption? So, I don't think we need a promise there.

That leaves us with needing some kind of solution for `openApp()` and I think I would be mostly in favor of adding a single `Onyx.connect()` inside of `openApp()` which isn't exposed to anything else.

Reply:
>Would initialKeysToCache be different than the existing param for initialKeyStates?

I think they're different as `initialKeyStates` is used to set a default for a key and not "pre-fill" the cache with already stored values.

>However, I'm not really sure that this would solve the problem that this PR is intending to solve, or if it does, I don't see how it solve the problem in this PR

I'm not sure either. Just a random idea that came to mind.

>It seems to me like the policies will definitely be read from Onyx when reconnectApp() is called. Is that a correct assumption?

Agree. I'm not seeing why we need this for `ReconnectApp`.

Reply:
> > It seems to me like the policies will definitely be read from Onyx when reconnectApp() is called. Is that a correct assumption?
> 
> Agree. I'm not seeing why we need this for `ReconnectApp`.

Oh, yes, you are right, the variables not having been read from Onyx in time is only a problem for `OpenApp`

Reply:
> That leaves us with needing some kind of solution for openApp() and I think I would be mostly in favor of adding a single Onyx.connect() inside of openApp() which isn't exposed to anything else.

This pattern is in a few places, it is almost like a `Onyx.get`, but async, right? I wonder if we should have a `Onyx.connectOneTime` or an extra parameter to `Onyx.connect` to handle the disconnection.

Reply:
I personally don't think we need any wrapper or functionality around it. I would probably think differently if I saw it misused or abused in several locations.